resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ResourceRequestHelper.scala

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.yarn
+
+import java.lang.{Long => JLong}
+import java.lang.reflect.InvocationTargetException
+
+import scala.collection.mutable
+import scala.util.Try
+
+import org.apache.hadoop.yarn.api.records.Resource
+
+import org.apache.spark.{SparkConf, SparkException}
+import org.apache.spark.deploy.yarn.config._
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config._
+import org.apache.spark.util.Utils
+
+/**
+ * This helper class uses some of Hadoop 3 methods from the YARN API,
+ * so we need to use reflection to avoid compile error when building against Hadoop 2.x
+ */
+private object ResourceRequestHelper extends Logging {
+  private val AMOUNT_AND_UNIT_REGEX = "([0-9]+)([A-Za-z]*)".r
+  private val RESOURCE_INFO_CLASS = "org.apache.hadoop.yarn.api.records.ResourceInformation"
+
+  /**
+   * Validates sparkConf and throws a SparkException if any of standard resources (memory or cores)
+   * is defined with the property spark.yarn.x.resource.y
+   */
+  def validateResources(sparkConf: SparkConf): Unit = {
+    val resourceDefinitions = Seq[(String, String)](
+      (AM_MEMORY.key, YARN_AM_RESOURCE_TYPES_PREFIX + "memory"),
```

Comment:
Still waiting for a word on this.

Reply:
Sure!
Did you mean this documentation? 
https://hadoop.apache.org/docs/r3.0.1/hadoop-yarn/hadoop-yarn-site/ResourceModel.html
I think it's required to check all the keys for memory / vcore that YARN deprecates, as those will flow trough Spark and eventually reach YARN's `ResourceInformation` and it will just blow up as only `memory-mb` and `vcores` are the ones that are not deprecated. The reason why it haven't caused a problem with current Spark code as it is using the `Resource` object and not using `ResourceInformation` at all.
So we need to disallow these:
- cpu-vcores
- memory
- mb

What do you think?

Reply:
I'm not familiar with the YARN code or what it does here.

I'm just worried about users setting cpu/memory resources outside of the proper Spark settings, and also the inconsistency in your code (using both memory and memory-mb).

Reply:
These are two separate things: 
1. One is that I don't reject all the deprecated standard resources has been known to YARN (explained in previous comment) which I will address soon.
2. Using `memory-mb` is the only way to initialize the memory resource with the YARN client, with the method `ResourceUtils.reinitializeResources`. 
I played around with this a bit, if I omit the standard resources and try to specify custom resources and then call `ResourceUtils.reinitializeResources`, an internal YARN exception will be thrown. 
Unfortunately, invoking this method is the most simple way to build tests upon custom resource types, to my best knowledge, so I can't really do much about this. 

> and also the inconsistency in your code (using both memory and memory-mb).
What did you mean with this? The only use of `"memory"` all around the change is to prevent it from being used with the new resource configs.

Reply:
> What did you mean with this?

I meant you were initializing `memory-mb` in tests but checking only `memory` here. That smells like you should be checking `memory-mb` here.

There kinds of things should have comments in the code so in the future we know why they are that way.

Reply:
Please see my last commit with the updates.
I only added some tests, so they are not extensive for every combination of spark resources and YARN standard resources. If you think I can add more testcases but I think this is fine as it is.

Reply:
Sure, adding some explanatory comments with my next commit.

Reply:
I think the code is now complete, please check!